### PR TITLE
Protect RPC interface against CSRF

### PR DESCRIPTION
**Pull request by kingosticks**
A fix for some of https://github.com/mopidy/mopidy/issues/1659

### CODE REVIEW

**Comment by kingosticks** [comment]
I will take a stab at the websocket too

File: mopidy/http/ext.conf [hunk 1]
```diff
@@ -4,3 +4,4 @@ hostname = 127.0.0.1
 port = 6680
 static_dir =
 zeroconf = Mopidy HTTP server on $hostname
+allowed_origins =
```
Comment (jodal) [inline mopidy/http/ext.conf]: Does `localhost:6600` work without whitelisting it, or should we make that the default?

File: docs/ext/http.rst [hunk 2]
```diff
@@ -98,3 +98,11 @@ See :ref:`config` for general help on configuring Mopidy.
     be published.
 
     Set to an empty string to disable Zeroconf for HTTP.
+
+.. confval:: http/allowed_origins
+
+    A whitelist of domains allowed to perform Cross-Origin Resource Sharing
+    (CORS) requests. Entries must be in the format ``hostname``:``port``.
```
Comment (jodal) [inline docs/ext/http.rst]: - [ ] Consider s/whitelist/list/ and explain that multiple values can be separated by comma or put on separate lines. I don't remember how well we explain how to use `config.List` config values elsewhere.
- [ ] Include the colon in the monospace part: ``hostname:port``.

File: mopidy/http/handlers.py [hunk 3]
```diff
@@ -19,12 +20,16 @@
 
 def make_mopidy_app_factory(apps, statics):
     def mopidy_app_factory(config, core):
+        origin_whitelist = {
```
Comment (jodal) [inline mopidy/http/handlers.py]: s/origin_whitelist/allowed_origins/ to keep the naming consistent?

File: mopidy/http/handlers.py [hunk 4]
```diff
@@ -143,16 +148,31 @@ def set_mopidy_headers(request_handler):
         'X-Mopidy-Version', mopidy.__version__.encode('utf-8'))
 
 
+def check_origin(origin, request_headers, origin_whitelist):
+    if origin is None:
+        logger.debug('Origin was not set')
+        return False
+    origin_whitelist.add(request_headers.get('Host', None))
```
Comment (jodal) [inline mopidy/http/handlers.py]: This allows access for a webpage at `foo:80` if the RPC interface also is available at `foo:80`? Cute. That voids my comment about `localhost:6600` as the default config value.

File: mopidy/http/handlers.py [hunk 5]
```diff
@@ -177,6 +197,18 @@ def set_extra_headers(self):
         self.set_header('Accept', 'application/json')
         self.set_header('Content-Type', 'application/json; utf-8')
 
+    def options(self):
+        origin = self.request.headers.get('Origin', None)
```
Comment (jodal) [inline mopidy/http/handlers.py]: `None` is the default return value for `.get()`, so you can leave it out.

**Comment by kingosticks** [comment]
So I think I have addressed those comments.

Now for the websocket, I thought we could just use the same function:
```
    def check_origin(self, origin):
        return check_origin(origin, self.request.headers, self.allowed_origins)
```

**Comment by kingosticks** [comment]
Fixed conflict and added a bit about same-origin requests to the docs and a changelog entry.

**Comment by jodal** [comment]
Thanks for taking the time to fixing this properly :-)

**Pull request merged**